File: engine/src/main/java/de/hanno/hpengine/engine/camera/Camera.java
package de.hanno.hpengine.engine.camera;

import de.hanno.hpengine.engine.config.Config;
import de.hanno.hpengine.engine.model.Entity;
import de.hanno.hpengine.engine.model.StaticModel;
import de.hanno.hpengine.log.ConsoleLogger;
import de.hanno.hpengine.util.Util;
import org.joml.Vector4f;
import org.lwjgl.BufferUtils;
import org.joml.Matrix4f;
import org.joml.Vector3f;

import java.io.IOException;
import java.io.ObjectOutputStream;
import java.nio.FloatBuffer;
import java.util.logging.Logger;

public class Camera extends Entity {
	private static Logger LOGGER = ConsoleLogger.getLogger();

	transient FloatBuffer viewProjectionMatrixBuffer = BufferUtils.createFloatBuffer(16);
	transient FloatBuffer projectionMatrixBuffer = BufferUtils.createFloatBuffer(16);
	transient FloatBuffer lastViewMatrixBuffer = BufferUtils.createFloatBuffer(16);

	protected Matrix4f projectionMatrix = new Matrix4f();

	protected Frustum frustum = new Frustum();

	private float near = 1f;
	private float far = 3000f;
	private float fov = 30f;
	private float ratio = (float) Config.getInstance().getWidth() / (float) Config.getInstance().getHeight();
	private float width = 1600f;
	private float height = 1600f;

	private boolean isPerspective = true;

	public Camera() {
		this.name = "Camera_" +  System.currentTimeMillis();
        init(Util.createPerspective(45f, (float) Config.getInstance().getWidth() / (float) Config.getInstance().getHeight(), this.near, this.far), this.near, this.far, 60f, (float) Config.getInstance().getWidth() / (float) Config.getInstance().getHeight());
		//this(renderer, Util.createOrthogonal(-1f, 1f, -1f, 1f, -1f, 2f), Util.lookAt(new Vector3f(1,10,1), new Vector3f(0,0,0), new Vector3f(0, 1, 0)));
	}
	public Camera(float near, float far, float fov, float ratio) {
		this.name = "Camera_" +  System.currentTimeMillis();
        init(Util.createPerspective(fov, ratio, near, far), near, far, fov, ratio);
	}

    public Camera(Camera camera) {
		this.name = "Camera_" +  System.currentTimeMillis();
        init(camera);
    }


	public Camera(Matrix4f projectionMatrix, float near, float far, float fov, float ratio) {
		this.name = "Camera_" +  System.currentTimeMillis();
        init(projectionMatrix, near, far, fov, ratio);
	}

    public void init(Camera camera) {
		set(camera);
        init(camera.getProjectionMatrix(), camera.getNear(), camera.getFar(), camera.getFov(), camera.getRatio());
        if(camera.hasParent()) {
            setParent(camera.getParent());
        }
    }

    public void init(Matrix4f projectionMatrix, float near, float far, float fov, float ratio) {
        this.near = near;
        this.far = far;
        this.fov = fov;
        this.ratio = ratio;
        this.projectionMatrix.set(projectionMatrix);

        frustum.calculate(this);
        saveViewMatrixAsLastViewMatrix();
        transform();
        storeMatrices();
    }

    public Camera(Vector3f position, String name, StaticModel model) {
		super(position, name, model);
        saveViewMatrixAsLastViewMatrix();
        transform();
        storeMatrices();
	}

	public void update(float seconds) {
		saveViewMatrixAsLastViewMatrix();
		super.update(seconds);
        if(hasMoved()) {
            transform();
            storeMatrices();
        }
	}

	private Matrix4f tempViewProjectionMatrix = new Matrix4f();
	private void storeMatrices() {
		synchronized (projectionMatrixBuffer) {
			projectionMatrixBuffer.rewind();
			projectionMatrix.get(projectionMatrixBuffer);
			projectionMatrixBuffer.rewind();
		}
		synchronized (viewProjectionMatrixBuffer) {
			viewProjectionMatrixBuffer.rewind();
			tempViewProjectionMatrix.set(projectionMatrix).mul(getViewMatrix()).get(viewProjectionMatrixBuffer);
			viewProjectionMatrixBuffer.rewind();
		}
	}

	private void transform() {
		frustum.calculate(this);
	}
	
	public Matrix4f getProjectionMatrix() {
		return projectionMatrix;
	}

	public void setProjectionMatrix(Matrix4f projectionMatrix) {
		this.projectionMatrix = projectionMatrix;
	}

	public FloatBuffer getLastViewMatrixAsBuffer() {
		return lastViewMatrixBuffer;
	}

	public void saveViewMatrixAsLastViewMatrix() {
		lastViewMatrixBuffer.rewind();
		lastViewMatrixBuffer.put(getViewMatrixAsBuffer(false));
		lastViewMatrixBuffer.rewind();
	}

	public FloatBuffer getProjectionMatrixAsBuffer() {
		return projectionMatrixBuffer.asReadOnlyBuffer();
	}

	public Frustum getFrustum() {
		return frustum;
	}

	public float getNear() {
		return near;
	}

	public void setNear(float near) {
		this.near = near;
		calculateProjectionMatrix();
		frustum.calculate(this);
	}


	public void setFar(float far) {
		this.far = far;
		calculateProjectionMatrix();
		frustum.calculate(this);
	}

	public float getFar() {
		return far;
	}

	public float getWidth() {
		return width;
	}

	public void setWidth(float width) {
		this.width = width;
        calculateProjectionMatrix();
        frustum.calculate(this);
	}

	public float getHeight() {
		return height;
	}

	public void setHeight(float height) {
		this.height = height;
        calculateProjectionMatrix();
        frustum.calculate(this);
	}

	public boolean isPerspective() {
		return isPerspective;
	}

	public void setPerspective(boolean isPerspective) {
		this.isPerspective = isPerspective;
	}

	private void calculateProjectionMatrix() {
		if(isPerspective) {
			projectionMatrix = Util.createPerspective(fov, ratio, near, far);
		} else {
			projectionMatrix = Util.createOrthogonal(-width/2, width/2, height/2, -height/2, -far/2, far/2);
		}
	}

	public void setRatio(float ratio) {
		this.ratio = ratio;
        calculateProjectionMatrix();
        frustum.calculate(this);
	}

	public void setFov(float fov) {
		this.fov = fov;
        calculateProjectionMatrix();
        frustum.calculate(this);
	}

	public float getFov() {
		return fov;
	}

	public float getRatio() {
		return ratio;
	}

	public FloatBuffer getViewProjectionMatrixAsBuffer() {
		return viewProjectionMatrixBuffer;
	}


	private void readObject(java.io.ObjectInputStream in) throws IOException, ClassNotFoundException {
		in.defaultReadObject();
		viewProjectionMatrixBuffer = BufferUtils.createFloatBuffer(16);
		projectionMatrixBuffer = BufferUtils.createFloatBuffer(16);
		lastViewMatrixBuffer = BufferUtils.createFloatBuffer(16);
	}


	private void writeObject(ObjectOutputStream oos) throws IOException {
		oos.defaultWriteObject();
	}

	public Vector3f[] getFrustumCorners() {
		Vector4f[] result = new Vector4f[8];
		Vector3f[] resultVec3 = new Vector3f[8];
		Matrix4f inverseProjection = getProjectionMatrix().invert(new Matrix4f());
		result[0] = inverseProjection.transform(new Vector4f(-1, 1, 1, 1));
		result[1] = inverseProjection.transform(new Vector4f(1, 1, 1, 1));
		result[2] = inverseProjection.transform(new Vector4f(1, -1, 1, 1));
		result[3] = inverseProjection.transform(new Vector4f(-1, -1, 1, 1));

		result[4] = inverseProjection.transform(new Vector4f(-1, 1, -1, 1));
		result[5] = inverseProjection.transform(new Vector4f(1, 1, -1, 1));
		result[6] = inverseProjection.transform(new Vector4f(1, -1, -1, 1));
		result[7] = inverseProjection.transform(new Vector4f(-1, -1, -1, 1));

		Matrix4f inverseView = getViewMatrix().invert(new Matrix4f());
		for(int i = 0; i < 8; i++) {
			result[i] = result[i].div(result[i].w);
			result[i] = inverseView.transform(result[i]);
			resultVec3[i] = new Vector3f(result[i].x, result[i].y, result[i].z);
		}

		return resultVec3;
	}
}

File: engine/src/main/java/de/hanno/hpengine/engine/camera/Camera.kt
package de.hanno.hpengine.engine.camera

import de.hanno.hpengine.engine.component.Component
import de.hanno.hpengine.engine.entity.Entity
import de.hanno.hpengine.engine.graphics.shader.safePut
import de.hanno.hpengine.engine.scene.Scene
import de.hanno.hpengine.log.ConsoleLogger
import de.hanno.hpengine.util.Util
import org.joml.AxisAngle4f
import org.joml.Matrix4f
import org.joml.Quaternionf
import org.joml.Vector3f
import org.joml.Vector4f
import org.lwjgl.BufferUtils
import java.nio.FloatBuffer

open class Camera @JvmOverloads constructor(
        override val entity: Entity,
        ratio: Float = 1280f/720f): Component {

    var exposure = 5f
    var focalDepth = Defaults.focalDepth
    var focalLength = Defaults.focalLength
    var fStop = Defaults.fStop

    var ratio = ratio
        set(value) {
            field = value
            updateProjectionMatrixAndFrustum()
        }
    var viewProjectionMatrixAsBuffer = BufferUtils.createFloatBuffer(16)
    var projectionMatrixAsBuffer = BufferUtils.createFloatBuffer(16)
    var lastViewMatrixAsBuffer = BufferUtils.createFloatBuffer(16)
    val viewMatrixAsBuffer = BufferUtils.createFloatBuffer(16)

    var viewMatrix = Matrix4f()
        get() {
            return entity.transform.transformation.invert(field)
        }
    var projectionMatrix = Matrix4f()
    var viewProjectionMatrix = Matrix4f()

    var frustum = Frustum()

    var near = 0.1f
        set(value) {
            field = value
            updateProjectionMatrixAndFrustum()
        }
    var far = 2000f
        set(value) {
            field = value
            updateProjectionMatrixAndFrustum()
        }

    var fov = Defaults.fov
        set(value) {
            field = value
            updateProjectionMatrixAndFrustum()
        }

    var width = 1600f
        set(width) {
            field = width
            updateProjectionMatrixAndFrustum()
        }
    var height = 1600f
        set(height) {
            field = height
            updateProjectionMatrixAndFrustum()
        }

    var perspective = true

    val frustumCorners: Array<Vector3f>
        get() {
            val inverseProjection = projectionMatrix.invert(Matrix4f())
            val result = mutableListOf<Vector4f>(inverseProjection.transform(Vector4f(-1f, 1f, 1f, 1f)),
                    inverseProjection.transform(Vector4f(1f, 1f, 1f, 1f)),
                    inverseProjection.transform(Vector4f(1f, -1f, 1f, 1f)),
                    inverseProjection.transform(Vector4f(-1f, -1f, 1f, 1f)),
                    inverseProjection.transform(Vector4f(-1f, 1f, -1f, 1f)),
                    inverseProjection.transform(Vector4f(1f, 1f, -1f, 1f)),
                    inverseProjection.transform(Vector4f(1f, -1f, -1f, 1f)),
                    inverseProjection.transform(Vector4f(-1f, -1f, -1f, 1f)))

            val inverseView = viewMatrix.invert(Matrix4f())
            val resultVec3 = result.map { it.div(it.w) }.map { inverseView.transform(it) }.map { Vector3f(it.x, it.y, it.z) }

            return resultVec3.toTypedArray()
        }

    init {
        init(Util.createPerspective(45f, ratio, this.near, this.far), this.near, this.far, 60f, ratio, 5f, Defaults.focalDepth, Defaults.focalLength, Defaults.fStop)
    }

    constructor(entity: Entity, near: Float, far: Float, fov: Float, ratio: Float): this(entity) {
        init(Util.createPerspective(fov, ratio, near, far), near, far, fov, ratio, 5f, Defaults.focalDepth, Defaults.focalLength, Defaults.fStop)
    }

    constructor(entity: Entity, camera: Camera): this(entity) {
        init(camera)
    }

    constructor(entity: Entity, projectionMatrix: Matrix4f, near: Float, far: Float, fov: Float, ratio: Float): this(entity) {
        init(projectionMatrix, near, far, fov, ratio, 5f, Defaults.focalDepth, Defaults.focalLength, Defaults.fStop)
    }

    fun init(camera: Camera) {
        entity.transform.set(camera.entity.transform)
        init(camera.projectionMatrix, camera.near, camera.far,
                camera.fov, camera.ratio, camera.exposure,
                camera.focalDepth, camera.focalLength, camera.fStop)

        if (camera.entity.hasParent) {
            val formerParent = camera.entity.parent
            entity.removeParent()
            entity.parent = formerParent
        }
    }

    fun init(projectionMatrix: Matrix4f, near: Float, far: Float, fov: Float, ratio: Float,
             exposure: Float, focalDepth: Float, focalLength: Float, fStop: Float) {

        this.exposure = exposure
        this.focalDepth = focalDepth
        this.focalLength = focalLength
        this.fStop = fStop

        this.near = near
        this.far = far
        this.fov = fov
        this.ratio = ratio
        this.projectionMatrix.set(projectionMatrix)

        saveViewMatrixAsLastViewMatrix()
        transform()
        storeMatrices()
    }

    override suspend fun update(scene: Scene, deltaSeconds: Float) {
        this@Camera.saveViewMatrixAsLastViewMatrix()
        this@Camera.projectionMatrix.mul(this@Camera.viewMatrix, this@Camera.viewProjectionMatrix) // TODO: Should move into the block below, but it's currently broken
        this@Camera.frustum.frustumIntersection.set(this@Camera.viewProjectionMatrix)
//        TODO: Fix this, doesn't work
//        if (entity.hasMoved())
        this@Camera.run {
            this.transform()
            this.storeMatrices()
        }
    }

    private fun storeMatrices() {

        viewMatrixAsBuffer.rewind()
        viewMatrix.get(viewMatrixAsBuffer)
        viewMatrixAsBuffer.rewind()

        projectionMatrixAsBuffer.rewind()
        projectionMatrix.get(projectionMatrixAsBuffer)
        projectionMatrixAsBuffer.rewind()

        viewProjectionMatrixAsBuffer.rewind()
        viewProjectionMatrix.set(projectionMatrix).mul(viewMatrix).get(viewProjectionMatrixAsBuffer)
        viewProjectionMatrixAsBuffer.rewind()
    }

    private fun transform() {
        frustum.calculate(this)
    }

    fun saveViewMatrixAsLastViewMatrix() = lastViewMatrixAsBuffer.safePut(viewMatrixAsBuffer)

    private fun updateProjectionMatrixAndFrustum() {
        calculateProjectionMatrix()
        frustum.calculate(this)
    }

    private fun calculateProjectionMatrix() {
        if (perspective) {
            projectionMatrix = Util.createPerspective(fov, ratio, near, far)
        } else {
            projectionMatrix = Util.createOrthogonal(-width / 2, width / 2, height / 2, -height / 2, -far / 2, far / 2)
        }
    }


    fun getViewDirection() = entity.transform.viewDirection
    fun getRightDirection() = entity.transform.rightDirection
    fun getUpDirection() = entity.transform.upDirection
    fun rotation(quaternion: Quaternionf) = entity.transform.rotation(quaternion)
    fun rotate(axisAngle4f: AxisAngle4f) = entity.transform.rotate(axisAngle4f)
    fun getTranslation(dest: Vector3f) = entity.transform.getTranslation(dest)
    fun translateLocal(offset: Vector3f) = entity.transform.translateLocal(offset)
    fun setTranslation(translation: Vector3f) = entity.transform.setTranslation(translation)
    fun getPosition() = entity.transform.position


    fun getTranslationRotationBuffer(): FloatBuffer {
        return viewMatrixAsBuffer
    }

    companion object {
        private val LOGGER = ConsoleLogger.getLogger()
    }

    object Defaults {
        const val focalDepth = 1.84f
        const val focalLength = 51f
        const val fStop = 1.15f
        const val fov = 60f
    }
}


